### PR TITLE
nwchem: migrate to python@3.12

### PR DESCRIPTION
**Pull request by edoapra**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

**Comment by p-linnane** [comment]
Is the patch for Python 3.12 compatibility?

**Comment by edoapra** [comment]
> Is the patch for Python 3.12 compatibility?

Yes, it is. The buggy code was already there for Python versions < 3.12. However, it generates a segfault  with 3.12.
The patch addresses this bug.

File: Formula/n/nwchem.rb [hunk 1]
```diff
@@ -25,11 +25,16 @@ class Nwchem < Formula
...
   depends_on "openblas"
-  depends_on "python@3.11"
+  depends_on "python@3.12"
   depends_on "scalapack"
 
   uses_from_macos "libxcrypt"
 
+  patch do
+    url "https://github.com/nwchemgit/nwchem/commit/48fac057df694267c2422adc2b394a0ac0815c02.patch"
+    sha256 "4ed952f52256b7288f19b03451ead2fceaf1e54a4a499e8ced1f582d5217923c"
```
Comment (p-linnane) [inline Formula/n/nwchem.rb:35]:
```suggestion
  # Adds Python 3.12 compatibility.
  # Remove in next release.
  patch do
    url "https://github.com/nwchemgit/nwchem/commit/48fac057df694267c2422adc2b394a0ac0815c02.patch?full_index=1"
    sha256 "5514e33185ca34c099d26806112b08c582a5f79e000184dfd1b8c9dfdd5cc1d9"
```

**Comment by p-linnane** [comment]
Running `brew style --fix` should resolve this last issue.

File: Formula/n/nwchem.rb [hunk 2]
```diff
@@ -65,7 +72,7 @@ def install
       ENV.append "OMPI_CFLAGS", "-Wl,-ld_classic" if DevelopmentTools.clang_build_version >= 1500
       os = OS.mac? ? "MACX64" : "LINUX64"
       system "make", "nwchem_config", "NWCHEM_MODULES=all python gwmol", "USE_MPI=Y"
-      system "make", "NWCHEM_TARGET=#{os}", "USE_MPI=Y"
+      system "make", "NWCHEM_TARGET=#{os}", "USE_MPI=Y", "-j3"
```
Comment (branchvincent) [inline Formula/n/nwchem.rb:75]: brew already sets this via `MAKEFLAGS`: https://github.com/Homebrew/brew/blob/e57a87cfe9c15f192cd4fc6695a2b3b7acf0ddf7/Library/Homebrew/extend/ENV/std.rb#L38

Comment (edoapra) [reply Formula/n/nwchem.rb]: `-j3` removed

**Comment by edoapra** [comment]
I think the label "CI-long-timeout"/"long build" is needed again

**Review by chenrui333 — APPROVED** [review]

**Pull request merged**